# Credit Card Fraud Detection


Import The Libraries

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

Load The Dataset

In [ ]:
credit_data=pd.read_csv("/content/creditcard.csv")
credit_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
credit_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

Check The Null values

In [ ]:
credit_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Check The Legit Transction &  Fraulent Transction



0 -->  Normal/Legit  Transction



1  -->  Fraudalent Transction

In [ ]:
credit_data["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [ ]:
legit=credit_data[credit_data.Class==0]
fraud=credit_data[credit_data.Class==1]

In [ ]:
print(legit.shape)

(284315, 31)


In [ ]:
print(fraud.shape)

(492, 31)


In [ ]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

Compare The Value Both Transction

In [ ]:
credit_data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Bulit a  sample  dataset  which  containing  similar  distribution  of   Legit  Transaction  &   Fraudulent  Transaction



In 284315  Legit Transaction, We going  to  take  492  Transaction   from  Legit   Transaction

In [ ]:
legit_sample=legit.sample(n=492)
legit_sample.head()

Time        V1        V2        V3        V4        V5        V6  \
218138  141183.0 -0.725338  1.027874 -0.000308 -0.714915  0.135060 -0.980560   
34872    37863.0 -1.093749  1.021302  1.263211 -0.967866 -0.122500  0.554127   
140177   83581.0  1.279509  0.579254 -0.466759  0.835882  0.030951 -1.416235   
81745    59069.0 -0.381183  1.167640  1.527367  0.490555  0.200696 -0.756988   
248550  153995.0 -0.648921 -0.257187  1.107286 -0.587137  0.099952  0.088285   

              V7        V8        V9  ...       V21       V22       V23  \
218138  0.528476  0.376229 -0.288839  ... -0.100815 -0.380899  0.120510   
34872  -0.842832 -2.479357  0.180269  ... -1.229363  0.818690 -0.138811   
140177  0.376399 -0.300744 -0.114591  ... -0.141737 -0.366498 -0.076212   
81745   0.808207 -0.223698 -0.651154  ...  0.051977  0.154561 -0.185915   
248550 -0.362681  0.291613  0.818837  ...  0.187733  0.700566  0.553618   

             V24       V25       V26       V27       V28  Amount  Class  
218138  0.045496 -0.985025  0.039567  0.140514  0.164255    4.95      0  
34872  -0.515225 -0.070506  1.033592  0.278927  0.200193   15.99      0  
140177  0.285173  0.562092  0.369985 -0.035893  0.035972    0.76      0  
81745   0.412364  0.140134 -0.497011 -0.078139  0.034925    2.49      0  
248550  0.759629 -1.697999  0.074188  0.068476  0.283336    9.99      0  

[5 rows x 31 columns]

We   have  492  Legit   Transcation  and   492  Fraudlant  Transaction

In [ ]:
new_credit_data=pd.concat([legit_sample,fraud],axis=0)
new_credit_data.head()

Time        V1        V2        V3        V4        V5        V6  \
218138  141183.0 -0.725338  1.027874 -0.000308 -0.714915  0.135060 -0.980560   
34872    37863.0 -1.093749  1.021302  1.263211 -0.967866 -0.122500  0.554127   
140177   83581.0  1.279509  0.579254 -0.466759  0.835882  0.030951 -1.416235   
81745    59069.0 -0.381183  1.167640  1.527367  0.490555  0.200696 -0.756988   
248550  153995.0 -0.648921 -0.257187  1.107286 -0.587137  0.099952  0.088285   

              V7        V8        V9  ...       V21       V22       V23  \
218138  0.528476  0.376229 -0.288839  ... -0.100815 -0.380899  0.120510   
34872  -0.842832 -2.479357  0.180269  ... -1.229363  0.818690 -0.138811   
140177  0.376399 -0.300744 -0.114591  ... -0.141737 -0.366498 -0.076212   
81745   0.808207 -0.223698 -0.651154  ...  0.051977  0.154561 -0.185915   
248550 -0.362681  0.291613  0.818837  ...  0.187733  0.700566  0.553618   

             V24       V25       V26       V27       V28  Amount  Class  
218138  0.045496 -0.985025  0.039567  0.140514  0.164255    4.95      0  
34872  -0.515225 -0.070506  1.033592  0.278927  0.200193   15.99      0  
140177  0.285173  0.562092  0.369985 -0.035893  0.035972    0.76      0  
81745   0.412364  0.140134 -0.497011 -0.078139  0.034925    2.49      0  
248550  0.759629 -1.697999  0.074188  0.068476  0.283336    9.99      0  

[5 rows x 31 columns]

Dividing The Dataset into This Independant & Dependant variable

In [ ]:
x=new_credit_data.drop(["Class"],axis=1)
y=new_credit_data["Class"]

Split Data into Training and Testing Data

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
print(X_train.shape,y_train.shape,X_test.shape,y_test.shape)

(787, 30) (787,) (197, 30) (197,)


Applying  Logistic   Regression

In [ ]:
model=LogisticRegression(solver='lbfgs',max_iter=1000)
model.fit(X_train,y_train)

LogisticRegression(max_iter=1000)

In [ ]:
y_pred=model.predict(X_test)

In [ ]:
model_accuracy_score=accuracy_score(y_test,y_pred)

Check  Accuracy  score

In [ ]:
print(model_accuracy_score)

0.9593908629441624


Check The Accuracy, Recall, Precision and F1-score

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.93      1.00      0.96        99
           1       1.00      0.92      0.96        98

    accuracy                           0.96       197
   macro avg       0.96      0.96      0.96       197
weighted avg       0.96      0.96      0.96       197

